cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/connector/ConnectorSpec.java

```diff
@@ -17,20 +17,39 @@
 
 package io.cdap.cdap.etl.api.connector;
 
+import io.cdap.cdap.api.data.schema.Schema;
+
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.Map;
 import java.util.Objects;
+import java.util.Set;
+import javax.annotation.Nullable;
 
 /**
  * The connector spec contains all the properties based on the path and plugin config
  */
 public class ConnectorSpec {
+  // schema is null when the connector is unable to retrieve it from the resource
+  private final Schema schema;
+  private final Set<PluginSpec> relatedPlugins;
```

Comment:
separating this from the properties means the same set of properties applies to all plugins. This can be weird if the plugins require different properties, or even impossible if multiple plugins need to configure the same property in different ways. It would be better if the properties, type, and name were grouped together in the same class.